---
title: "R Assignment Script"
author: "Kofi Antwi Appiagyei"
date: "2025-03-21"
output:
  pdf_document: default
  output_file: "R_Assignment_Script.pdf"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Below is my workflow for the R Assignment
Load libraries
```{r, echo=TRUE}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyverse)
library(tidyr)
library(purrr)

```


DATA INSPECTION


# Read the files

```{r, echo=TRUE}


genotypes <- read.table("fang_et_al_genotypes.txt", header = TRUE, sep = "\t", stringsAsFactors = FALSE)
snp_position <- read.table("snp_position.txt", header = TRUE, sep = "\t", stringsAsFactors = FALSE)

```

Inspecting fang_et_al_genotypes.txt
1. File size (in bytes)

```{r, echo=TRUE}

file_size <- file.info("fang_et_al_genotypes.txt")$size
print(paste("File size (bytes):", file_size))

```

2. View the first 6 rows (head) of the data

```{r, echo=TRUE}

head(genotypes[, 1:8])

```

3. View the last 6 rows (tail) of the data
```{r, echo=TRUE}

tail(genotypes[, 1:8])

```

4. Number of rows and columns in the data
```{r, echo=TRUE}

num_rows <- nrow(genotypes)
num_cols <- ncol(genotypes)
print(paste("Number of rows:", num_rows))
print(paste("Number of columns:", num_cols))

```

5. Check for missing data in the data
Check for the presence of "?" in the entire dataset

```{r, echo=TRUE}

missing_data_placeholder <- sum(genotypes == "?/?")
print(paste("Number of '?/?' placeholders:", missing_data_placeholder))

```

6. Most common element in the "Group" column

```{r, echo=TRUE}

most_common_group <- names(sort(table(genotypes$Group), decreasing = TRUE))[1]
print(paste("Most common element in Group column:", most_common_group))

```

7. Most common element in the "Gene" column

```{r, echo=TRUE}

most_common_gene <- names(sort(table(snp_position$gene), decreasing = TRUE))[1]
print(paste("Most common element in gene column:", most_common_gene))
```

# Observations from Data Inspection
- After inspection, I learnt the file size is 10.54 mb
- The data has 2782 rows and 986 columns
- There is 135452 missing data encoded by ?/?
- The most common group is "ZMMLR"
- The file is ASCII text with very long lines


```{r}
#Inspecting snp_position.txt
# Read the file
snp_position <- read.table("snp_position.txt", header = TRUE, sep = "\t", stringsAsFactors = FALSE)

```

After inspection, I learnt the file size is 10.54 mb
The data has 2782 rows and 986 columns
There is 135452 missing data encoded by ?/?
The most common group is "ZMMLR"
The file is ASCII text with very long lines


Inspecting snp_position.txt

1. File size (in bytes)
```{r, echo=TRUE}


file_size_snp <- file.info("snp_position.txt")$size
print(paste("File size (bytes):", file_size_snp))

```

2. View the first 6 rows (head) of the data

```{r, echo=TRUE}

head(snp_position)

```

3. View the last 6 rows (tail) of the data

```{r, echo=TRUE}

tail(snp_position)

```

4. Number of rows and columns in the data

```{r, echo=TRUE}

num_rows_snp <- nrow(snp_position)
num_cols_snp <- ncol(snp_position)
print(paste("Number of rows:", num_rows_snp))
print(paste("Number of columns:", num_cols_snp))

```


5. Check for missing data in the data
Check for the presence of "?" in the entire dataset

```{r, echo=TRUE}

missing_data_placeholder <- sum(snp_position == "?/?")
print(paste("Number of '?/?' placeholders:", missing_data_placeholder))

```

6. Most common element in the "Chromosome" column

```{r, echo=TRUE}

most_common_chromosome <- names(sort(table(snp_position$Chromosome), decreasing = TRUE))[1]
print(paste("Most common element in Chromosome column:", most_common_chromosome))
```

```



After inspection, I learnt the file size is 79 kb
The data has 983 rows and 15 columns
There is no missing data
The most common chromosome number is 1
The most common gene type is "Zmm28"
The file is ASCII text with very long lines



DATA PROCESSING

Transpose the data


```{r, echo=TRUE}

transposed_genotypes <- as.data.frame(t(genotypes), stringsAsFactors = FALSE)


```

Convert first row to column names

```{r, echo=TRUE}

colnames(transposed_genotypes) <- transposed_genotypes[3, ]  

```


Remove the first row as it's now the column names

```{r, echo=TRUE}

transposed_genotypes <- transposed_genotypes[-c(1,2) ]

```

Add original column names as a new first column

```{r, echo=TRUE}

transposed_genotypes <- cbind(Original_Colnames = rownames(transposed_genotypes), transposed_genotypes)
transposed_genotypes <- transposed_genotypes[-c(1:3), ]
colnames(transposed_genotypes)[1] <- "SNP_ID"

```


Reset row names

```{r, echo=TRUE}

rownames(transposed_genotypes) <- NULL

```


Extract needed columns for merging

```{r, echo=TRUE}

snp_extract <- select(snp_position, SNP_ID, Chromosome, Position)

```


Merge snp_position and transposed_genotypes by "SNP_ID"

```{r, echo=TRUE}

colnames(transposed_genotypes) <- make.unique(colnames(transposed_genotypes))
merged <- left_join(snp_extract, transposed_genotypes, by = "SNP_ID")

```


Extract maize columns

```{r, echo=TRUE}

maize <- merged %>% select(SNP_ID, Chromosome = Chromosome, Position = Position, 
                                 starts_with("ZMMIL"), starts_with("ZMMLR"), starts_with("ZMMMR"))

```


Subset data by specific values in the Chromosome column and sort by Position

```{r, echo=TRUE}

subset_data <- function(data, value, filename) {

# Subset the data for the given Chromosome value

  selected <- data[data$Chromosome == value, ]
  
  # Convert the Position column to numeric (ensuring it is numeric before sorting)
  selected$Position <- suppressWarnings(as.numeric(as.character(selected$Position)))
  
  # Sort the selected data by the Position column in ascending order
  selected_sorted <- selected[order(selected$Position), ]
  
  # Write the sorted data to the file
  write.table(selected_sorted, filename, sep = "\t", row.names = FALSE, quote = FALSE)
}

```

Loop through chromosomes 1 to 10 and subset

```{r, echo=TRUE}

for (i in 1:10) {
  subset_data(maize, i, paste0("Maize_chr", i, ".txt"))
}

```


Create a directory for ascending files and move them

```{r, echo=TRUE}

dir.create("Maize_ascend")
file.rename(list.files(pattern = "Maize_chr[0-9]+.txt"), file.path("Maize_ascend", list.files(pattern = "Maize_chr[0-9]+.txt")))

```


Extract the first line (header) from maize

```{r, echo=TRUE}

header <- head(maize, 1)

```

Creating multiple and unknown  chromosome files

```{r, echo=TRUE}

# Filter rows where the second column is "multiple" (for Maize_chrm.txt)
maize_chrm <- maize[maize$Chromosome == "multiple", ]

# Combine header with the filtered data and write to Maize_chrm.txt
maize_chrm_final <- rbind(header, maize_chrm)
write.table(maize_chrm_final, "Maize_chrm.txt", sep = "\t", row.names = FALSE, col.names = TRUE, quote = FALSE)

# Filter rows where the second column is "unknown" (for Maize_chru.txt)
maize_chru <- maize[maize$Chromosome == "unknown", ]

# Combine header with the filtered data and write to Maize_chru.txt
maize_chru_final <- rbind(header, maize_chru)
write.table(maize_chru_final, "Maize_chru.txt", sep = "\t", row.names = FALSE, col.names = TRUE, quote = FALSE)

```


replace ? with -

```{r, echo=TRUE}

maize_hyphen <- maize
maize_hyphen[maize_hyphen == "?/?"] <- "-/-"

```


Subset data by specific values in the Chromosome column and sort by Position in descending order

```{r, echo=TRUE}

subset_data <- function(data, value, filename) {
  # Subset the data for the given Chromosome value
  selected <- data[data$Chromosome == value, ]
  
  # Convert the Position column to numeric (ensure it is numeric before sorting)
  selected$Position <- suppressWarnings(as.numeric(as.character(selected$Position)))
  
  # Sort the selected data by the Position column in descending order
  selected_sorted <- selected[order(selected$Position, decreasing = TRUE), ]
  
  # Write the sorted data to the file
  write.table(selected_sorted, filename, sep = "\t", row.names = FALSE, quote = FALSE)
}

```

Loop through chromosomes 1 to 10 and subset

```{r, echo=TRUE}

for (i in 1:10) {
  subset_data(maize_hyphen, i, paste0("Maize_chrd", i, ".txt"))
}

```


Create a directory for descending files and move them

```{r, echo=TRUE}

dir.create("Maize_descend")
file.rename(list.files(pattern = "Maize_chrd[0-9]+.txt"), file.path("Maize_descend", list.files(pattern = "Maize_chrd[0-9]+.txt")))

```


FOR TEOSINTE

Extract teosinte columns
```{r, echo=TRUE}


teosinte <- merged %>% select(SNP_ID, Chromosome = Chromosome, Position = Position, 
                                 starts_with("ZMPBA"), starts_with("ZMPIL"), starts_with("ZMPJA"))

```


Subset data by specific values in the Chromosome column and sort by Position

```{r, echo=TRUE}

subset_data <- function(data, value, filename) {
  # Subset the data for the given Chromosome value
  selected <- data[data$Chromosome == value, ]
  
  # Convert the Position column to numeric (ensuring it is numeric before sorting)
  selected$Position <- suppressWarnings(as.numeric(as.character(selected$Position)))
  
  # Sort the selected data by the Position column in ascending order
  selected_sorted <- selected[order(selected$Position), ]
  
  # Write the sorted data to the file
  write.table(selected_sorted, filename, sep = "\t", row.names = FALSE, quote = FALSE)
}

```

Loop through chromosomes 1 to 10 and subset

```{r, echo=TRUE}

for (i in 1:10) {
  subset_data(teosinte, i, paste0("Teosinte_chr", i, ".txt"))
}

```


Create a directory for ascending files and move them

```{r, echo=TRUE}

dir.create("Teosinte_ascend")
file.rename(list.files(pattern = "Teosinte_chr[0-9]+.txt"), file.path("Teosinte_ascend", list.files(pattern = "Teosinte_chr[0-9]+.txt")))

```

Creating multiple and unknown chromosome files

```{r, echo=TRUE}

#Extract the first line (header) from teosinte
header <- head(teosinte, 1)

# Filter rows where the second column is "multiple" 
teosinte_chrm <- teosinte[teosinte$Chromosome == "multiple", ]

# Combine header with the filtered data and write to Teosinte_chrm.txt
teosinte_chrm_final <- rbind(header, teosinte_chrm)
write.table(teosinte_chrm_final, "Teosinte_chrm.txt", sep = "\t", row.names = FALSE, col.names = TRUE, quote = FALSE)

# Filter rows where the second column is "unknown" 
teosinte_chru <- teosinte[teosinte$Chromosome == "unknown", ]

# Combine header with the filtered data and write to Teosinte_chru.txt
teosinte_chru_final <- rbind(header, teosinte_chru)
write.table(teosinte_chru_final, "Teosinte_chru.txt", sep = "\t", row.names = FALSE, col.names = TRUE, quote = FALSE)


```

replace ? with -

```{r, echo=TRUE}

teosinte_hyphen <- teosinte
teosinte_hyphen[teosinte_hyphen == "?/?"] <- "-/-"

```

Subset data by specific values in the Chromosome column and sort by Position in descending order

```{r, echo=TRUE}

subset_data <- function(data, value, filename) {
  # Subset the data for the given Chromosome value
  selected <- data[data$Chromosome == value, ]
  
  # Convert the Position column to numeric 
  selected$Position <- suppressWarnings(as.numeric(as.character(selected$Position)))
  
  # Sort the selected data by the Position column in descending order
  selected_sorted <- selected[order(selected$Position, decreasing = TRUE), ]
  
  # Write the sorted data to the file
  write.table(selected_sorted, filename, sep = "\t", row.names = FALSE, quote = FALSE)
}

```

Loop through chromosomes 1 to 10 and subset

```{r, echo=TRUE}

for (i in 1:10) {
  subset_data(teosinte_hyphen, i, paste0("Teosinte_chrd", i, ".txt"))
}

```


Create a directory for descending files and move them

```{r, echo=TRUE}

dir.create("Teosinte_descend")
file.rename(list.files(pattern = "Teosinte_chrd[0-9]+.txt"), file.path("Teosinte_descend", list.files(pattern = "Teosinte_chrd[0-9]+.txt")))


```



DATA VISUALIZATION

Distribution of SNPs between chromosomes



Bar chart of the distribution of SNPs by Chromosome

```{r, echo=TRUE}

ggplot(data = maize) + 
    geom_bar(mapping = aes(x = Chromosome, fill = Chromosome)) +
  scale_x_discrete(limits = c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "multiple", "unknown")) +
  ggtitle("Distribution of SNPs by Chromosome in Maize") +
  theme(plot.title = element_text(hjust = 0.5)) +
  theme_classic()

```


Scatter plot of the distribution of SNPs by Chromosome

```{r, echo=TRUE}

ggplot(data = maize) + 
  geom_point(mapping = aes(x = Chromosome, y = Position, color = Chromosome)) + 
  scale_x_discrete(limits = c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "multiple", "unknown")) +
  ggtitle("Distribution of SNPs by Chromosome in Maize") +
  theme(plot.title = element_text(hjust = 0.5))

```

Heat map of SNPs by Chromosome

```{r, echo=TRUE}

ggplot(data = maize) + 
    geom_bin2d(mapping = aes(x = Chromosome, y = Position), bins = 30) +  # Use geom_bin2d for 2D binning
    scale_x_discrete(limits = c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "multiple", "unknown")) +
    ggtitle("Heat Density Plot of SNPs by Chromosome in Maize") +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme_classic() +
    scale_fill_viridis_c() 


```


# Plot the distribution of SNPs by Chromosome

FOR TEOSINTE

Bar chart of the distribution of SNPs by Chromosome

```{r, echo=TRUE}


ggplot(data = teosinte) + 
    geom_bar(mapping = aes(x = Chromosome, fill = Chromosome)) +
  scale_x_discrete(limits = c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "multiple", "unknown")) +
  ggtitle("Distribution of SNPs by Chromosome in Teosinte") +
  theme(plot.title = element_text(hjust = 0.5)) +
  theme_classic()

```


Scatter plot of the distribution of SNPs by Chromosome

```{r, echo=TRUE}

ggplot(data = teosinte) + 
  geom_point(mapping = aes(x = Chromosome, y = Position, color = Chromosome)) + 
  scale_x_discrete(limits = c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "multiple", "unknown")) +
  ggtitle("Distribution of SNPs by Chromosome in Teosinte") +
  theme(plot.title = element_text(hjust = 0.5))

```


Heat map of SNPs by Chromosome

```{r, echo=TRUE}

ggplot(data = teosinte) + 
    geom_bin2d(mapping = aes(x = Chromosome, y = Position), bins = 30) +  # Use geom_bin2d for 2D binning
    scale_x_discrete(limits = c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "multiple", "unknown")) +
    ggtitle("Heat Density Plot of SNPs by Chromosome in Teosinte") +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme_classic() +
    scale_fill_viridis_c() 
```


```{r, echo=TRUE}


# Count SNPs per chromosome for maize
maize_snp <- maize %>%
  group_by(Chromosome) %>%
  summarise(SNP_Count = n()) %>%
  mutate(Group = "Maize")

# Count SNPs per chromosome for teosinte
teosinte_snp <- teosinte %>%
  group_by(Chromosome) %>%
  summarise(SNP_Count = n()) %>%
  mutate(Group = "Teosinte")

# Combine data
snp_counts <- bind_rows(maize_snp, teosinte_snp)

```


Stacked Bar chart of SNP count per chromosome

```{r, echo=TRUE}

ggplot(snp_counts, aes(x = Chromosome, y = SNP_Count, fill = Group)) +
  geom_bar(stat = "identity", position = "dodge") +
  labs(title = "SNP Distribution Across Chromosomes in Maize and Teosinte",
       x = "Chromosome", y = "Number of SNPs") +
  scale_fill_manual(values = c("Maize" = "yellow", "Teosinte" = "purple")) + 
  scale_x_discrete(limits = c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "multiple", "unknown")) +
  theme_classic()

```

Scatter plot of merged data

```{r, echo=TRUE}

ggplot(snp_counts, aes(x = Chromosome, y = SNP_Count, color = Group)) +
  geom_point(size = 3, position = position_jitter(width = 0.2, height = 0)) + # Use 'geom_point' to create the scatter plot
  labs(title = "SNP Distribution Across Chromosomes in Maize and Teosinte",
       x = "Chromosome", y = "Number of SNPs") +
  scale_color_manual(values = c("Maize" = "yellow", "Teosinte" = "purple")) + 
  scale_x_discrete(limits = c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "multiple", "unknown")) + 
    theme_classic()

```


Heat map of merged data

```{r, echo=TRUE}

ggplot(snp_counts, aes(x = Chromosome, y = Group, fill = SNP_Count)) +
  geom_tile(color = "white ") +
  labs(title = "SNP Distribution Heatmap Across Chromosomes in Maize and Teosinte",
       x = "Chromosome", y = "Group", fill = "SNP Count") +
  scale_fill_gradient(low = "cyan", high = "blue") +
  scale_x_discrete(limits = c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "multiple", "unknown")) +
  theme_classic()


```
